File: 2023/csharp/Day2/Run/Logic.cs
namespace Day2;

public static class Logic
{
    public static int Part1(IEnumerable<string> lines) =>
        lines
            .AsParallel()
            .Select(line =>
            {
                var parts = line.Split(": ");
                var gameId = int.Parse(parts[0].Split().Last());

                var isValid = parts[1]
                    .Split("; ")
                    .All(set => set.Split(", ")
                        .All(c =>
                        {
                            var vec = c.Split();
                            var count = int.Parse(vec[0]);
                            var color = vec[1];
                            return count <= Data.MaxCubesLut[color];
                        }));

                return isValid ? gameId : 0;
            })
            .Sum();

    public static int Part2(IEnumerable<string> lines) =>
        lines
            .AsParallel()
            .Select(line => line.Split(": ")[1]
                .Split("; ")
                .SelectMany(set => set.Split(", "))
                .Select(cube =>
                {
                    var parts = cube.Split();
                    return (Color: parts[1], Count: int.Parse(parts[0]));
                })
                .GroupBy(c => c.Color)
                .Select(g => g.Max(c => c.Count))
                .Aggregate(1, (product, count) => product * count))
            .Sum();
}

File: 2023/rust/day2/src/logic.rs
use lazy_static::lazy_static;
use std::collections::HashMap;

pub fn part_1(lines: &[String]) -> u32 {
    lines
        .iter()
        .filter_map(|line| {
            let (game_info, game_data) = line.split_once(": ").unwrap();
            let game_id: u32 = game_info
                .split_whitespace()
                .last()
                .unwrap()
                .parse()
                .unwrap();

            game_data
                .split("; ")
                .all(|set| {
                    set.split(", ").all(|c| {
                        let vec = c.split_whitespace().collect::<Vec<_>>();
                        let count = vec[0].parse::<u32>().unwrap();
                        let color = vec[1];
                        count <= *MAX_CUBES_LUT.get(color).unwrap()
                    })
                })
                .then_some(game_id)
        })
        .sum()
}

pub fn part_2(lines: &[String]) -> u32 {
    lines
        .iter()
        .map(|line| {
            line.split(": ")
                .nth(1)
                .unwrap()
                .split("; ")
                .fold(HashMap::new(), |mut min_cubes, set| {
                    set.split(", ").for_each(|cube| {
                        let parts = cube.split_whitespace().collect::<Vec<_>>();
                        let count = parts[0].parse().unwrap();
                        let color = parts[1];
                        min_cubes
                            .entry(color)
                            .and_modify(|current| *current = u32::max(*current, count))
                            .or_insert(count);
                    });
                    min_cubes
                })
                .values()
                .product::<u32>()
        })
        .sum()
}

lazy_static! {
    pub static ref MAX_CUBES_LUT: HashMap<&'static str, u32> =
        HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);
}
